core/string/src/builder.rs

```diff
@@ -0,0 +1,423 @@
+/// s.extend([b'1', b'2', b'3']);
+/// let js_string = s.build();
+/// ```
+#[derive(Debug)]
+pub struct JsStringBuilder<T: private::JsStringData> {
+    cap: usize,
+    len: usize,
+    inner: NonNull<RawJsString>,
+    phantom_data: PhantomData<T>,
+}
+
+impl<D: private::JsStringData> Clone for JsStringBuilder<D> {
+    #[must_use]
+    fn clone(&self) -> Self {
+        let mut builder = Self::with_capacity(self.capacity());
+        // SAFETY:
+        // - `inner` must be a valid pointer, since it comes from a `NonNull`
+        // allocated above with the capacity of `s`, and initialize to `s.len()` in
+        // ptr::copy_to_non_overlapping below.
+        unsafe {
+            builder
+                .inner
+                .as_ptr()
+                .cast::<u8>()
+                .copy_from_nonoverlapping(self.inner.as_ptr().cast(), self.allocated_byte_len());
+
+            builder.set_len(self.len());
+        }
+        builder
+    }
+}
+
+impl<D: private::JsStringData> Default for JsStringBuilder<D> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<D: private::JsStringData> JsStringBuilder<D> {
+    const DATA_SIZE: usize = size_of::<D>();
+    const MIN_NON_ZERO_CAP: usize = 8 / Self::DATA_SIZE;
+
+    /// Create a new `JsStringBuilder` with capacity of zero.
+    #[inline]
+    #[must_use]
+    pub const fn new() -> Self {
+        Self {
+            cap: 0,
+            len: 0,
+            inner: NonNull::dangling(),
+            phantom_data: PhantomData,
+        }
+    }
+
+    /// Returns the number of elements that inner holds.
+    #[inline]
+    #[must_use]
+    pub const fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Forces the length of the [`JsStringBuilder`] to `new_len`.
+    ///
+    /// # Safety
+    ///
+    /// - `new_len` must be less than or equal to `capacity()`.
+    /// - The elements at `old_len..new_len` must be initialized.
+    ///
+    #[inline]
+    pub unsafe fn set_len(&mut self, new_len: usize) {
+        debug_assert!(new_len <= self.capacity());
+
+        self.len = new_len;
+    }
+
+    /// Returns the total number of elements can hold without reallocating
+    #[inline]
+    #[must_use]
+    pub const fn capacity(&self) -> usize {
+        self.cap
+    }
+
+    /// Returns the allocated byte of inner.
+    #[must_use]
+    const fn allocated_byte_len(&self) -> usize {
+        DATA_OFFSET + self.allocated_data_byte_len()
+    }
+
+    /// Returns the allocated byte of inner's data.
+    #[must_use]
+    const fn allocated_data_byte_len(&self) -> usize {
+        self.len() * Self::DATA_SIZE
+    }
+
+    /// Returns the capacity calculted from given layout.
+    #[must_use]
+    const fn capacity_from_layout(layout: Layout) -> usize {
+        (layout.size() - DATA_OFFSET) / Self::DATA_SIZE
+    }
+
+    /// create a new `JsStringBuilder` with specific capacity
+    #[must_use]
+    pub fn with_capacity(cap: usize) -> Self {
```

Comment:
We should add `#[inline]` on all public methods so the compiler can cross-crate inline them.